### PR TITLE
gNOI-2.1: Adding MTU config, ASIC loopback reflector mode and a new deviation

### PR DESCRIPTION
**Pull request by yashikakn**
1. Adding interface MTU configuration 
2. Setting EndpointType to ASIC Loopback mode in reflectorCreateRequest for Juniper 
3. Adding a new deviation to skip PLQ QualificationRateBytesPerSecond and ExpectedRateBytesPerSecond comparison

[06/27] Following changes done via commit '0c3ecbb'. Please refer to b/[282996826](https://urldefense.com/v3/__https:/issuetracker.google.com/issues/282996826__;!!NEt6yMaO-gk!GBsyQ3_fj-MTFgYPX900kBNFwQIf-Ejx6O-_A-wbIqdlT0xgVA5cYiOzZzMTy1E2-i_lRMDvf7d9cxhalNQI8Bo1h4s$) for details.
1. Removed QualificationRateBytesPerSecond and ExpectedRateBytesPerSecond comparison check  and the deviation 
2. Added check for Packets Sent/Received with a new deviation 
3. Edited Generator Post Sync duration to be zero

[06/28] Added ASIC loopback mode reflector capabilities check

### CODE REVIEW

**Review by sachendras — APPROVED** [review]

**Review by sachendras — DISMISSED** [review]

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go [hunk 1]
```diff
@@ -256,36 +278,74 @@ func TestLinkQualification(t *testing.T) {
...
-							Seconds: int64(plqDuration.preSyncDuration.Seconds()),
-						},
-						SetupDuration: &durationpb.Duration{
-							Seconds: int64(plqDuration.setupDuration.Seconds()),
-						},
-						PostSyncDuration: &durationpb.Duration{
-							Seconds: int64(plqDuration.postSyncDuration.Seconds()),
+	var reflectorCreateRequest *plqpb.CreateRequest
+	switch dut2.Vendor() {
+	case ondatra.JUNIPER:
```
Comment (greg-dennis) [inline feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:283]: Can you factor out the commonality between the cases? It looks like all cases have a single interface in the request that differs only in the endpoint type? If so, much of that logic can be common.

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go [hunk 2]
```diff
@@ -256,36 +278,74 @@ func TestLinkQualification(t *testing.T) {
...
-						PreSyncDuration: &durationpb.Duration{
-							Seconds: int64(plqDuration.preSyncDuration.Seconds()),
-						},
-						SetupDuration: &durationpb.Duration{
-							Seconds: int64(plqDuration.setupDuration.Seconds()),
-						},
-						PostSyncDuration: &durationpb.Duration{
-							Seconds: int64(plqDuration.postSyncDuration.Seconds()),
+	var reflectorCreateRequest *plqpb.CreateRequest
+	switch dut2.Vendor() {
```
Comment (greg-dennis) [inline feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:282]: Are we switching on the vendor without a deviation? Is there a discussion somewhere that says both types of requests are specification-compliant or are we awaiting resolution as to which one satisfies our expected contract?

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go (hunk 2 above)
Comment (yashikakn) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:282]: Both types of requests are specification-compliant. Please refer to : [https://github.com/openconfig/gnoi/blob/main/packet_link_qualification/index.md#the-service-implementation-must-support-multiple-generations-of-forwarding-hardware-which-have-different-capabilities-with-regards-to-both-generation-and-reflection-of-packets](url)

File: internal/deviations/deviations.go [hunk 3]
```diff
@@ -422,6 +422,11 @@ func RoutePolicyUnderAFIUnsupported(_ *ondatra.DUTDevice) bool {
 	return *routePolicyUnderAFIUnsupported
 }
 
+// SkipPlqQualificationRateCheck returns if PLQ QualificationRateBytesPerSecond and ExpectedRateBytesPerSecond comparison should be skipped
+func SkipPlqQualificationRateCheck(_ *ondatra.DUTDevice) bool {
```
[the remark this answers is not in the document]
Comment (yashikakn) [reply internal/deviations/deviations.go]: done. thanks.

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go (hunk 1 above)
Comment (yashikakn) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go]: Updated per your suggestion. Thanks.

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go [hunk 4]
```diff
@@ -191,6 +194,17 @@ func TestListDelete(t *testing.T) {
 	}
 }
 
+// configInterfaceMTU configures interface MTU.
```
Comment (greg-dennis) [inline feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:197]: I'd drop this comment. godoc comments are unnecessary on non-exported functions, and this comment is redundant just with the name.

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go [hunk 5]
```diff
@@ -200,6 +214,14 @@ func TestLinkQualification(t *testing.T) {
 	t.Logf("dut1: %v, dut2: %v", dut1.Name(), dut2.Name())
 	t.Logf("dut1 dp1 name: %v, dut2 dp2 name : %v", dp1.Name(), dp2.Name())
 
+	dut_list := []*ondatra.DUTDevice{dut1, dut2}
+	for _, dut := range dut_list {
```
Comment (greg-dennis) [inline feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go:218]: Go style is camelCase, not snake_case. In this case, I would just inline the dut list:
`for _, dut := range []*ondatra.DUTDevice{dut1, dut2} {`

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go (hunk 4 above)
Comment (yashikakn) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go]: Removed the comment. thanks.

File: internal/deviations/deviations.go [hunk 6]
```diff
@@ -428,6 +428,11 @@ func RoutePolicyUnderAFIUnsupported(_ *ondatra.DUTDevice) bool {
 	return *routePolicyUnderAFIUnsupported
 }
 
+// SkipPlqQualificationRateCheck returns if PLQ QualificationRateBytesPerSecond and ExpectedRateBytesPerSecond comparison should be skipped
```
[the remark this answers is not in the document]
Comment (yashikakn) [reply internal/deviations/deviations.go]: fixed it. thanks.

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go (hunk 5 above)
Comment (yashikakn) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go]: done. thanks.

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go [hunk 7]
```diff
@@ -256,36 +278,45 @@ func TestLinkQualification(t *testing.T) {
...
-				InterfaceName: dp2.Name(),
-				EndpointType: &plqpb.QualificationConfiguration_PmdLoopback{
-					PmdLoopback: &plqpb.PmdLoopbackConfiguration{},
+	intf := &plqpb.QualificationConfiguration{
+		Id:            plqID,
+		InterfaceName: dp2.Name(),
+
+		Timing: &plqpb.QualificationConfiguration_Rpc{
+			Rpc: &plqpb.RPCSyncedTiming{
+				Duration: &durationpb.Duration{
```
[the remark this answers is not in the document]
Comment (yashikakn) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go]: Fixed it. Thanks.

**Review by greg-dennis — APPROVED** [review]
Approving for Go quality

**Review by sachendras — APPROVED** [review]

**Comment by yashikakn** [comment]
@mojiiba , @singh-prem : Could you please review this pull req changes? Thanks.

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go [hunk 8]
```diff
@@ -113,6 +108,19 @@ func TestCapabilitiesResponse(t *testing.T) {
 			}
 		})
 	}
+
+	t.Run("Reflector", func(t *testing.T) {
+		if ((uint64(plqResp.GetReflector().GetAsicLoopback().GetMinSetupDuration().GetSeconds())) >= 1) &&
```
[the remark this answers is not in the document]
Comment (yashikakn) [reply feature/gnoi/packet_link_qualification/tests/packet_link_qualification_test/packet_link_qualification_test.go]: changes done. thanks.

**Review by greg-dennis — APPROVED** [review]

**Pull request merged**